frame/staking/src/lib.rs

```diff
@@ -1363,22 +1363,29 @@ decl_module! {
 			consumed_weight
 		}
 
-		/// Check if the current block number is the one at which the election window has been set
-		/// to open. If so, it runs the offchain worker code.
+		/// Offchain worker logic.
 		fn offchain_worker(now: T::BlockNumber) {
-			use offchain_election::{set_check_offchain_execution_status, compute_offchain_election};
-
-			if Self::era_election_status().is_open_at(now) {
-				let offchain_status = set_check_offchain_execution_status::<T>(now);
-				if let Err(why) = offchain_status {
-					log!(warn, "💸 skipping offchain worker in open election window due to [{}]", why);
-				} else {
-					if let Err(e) = compute_offchain_election::<T>() {
-						log!(error, "💸 Error in election offchain worker: {:?}", e);
-					} else {
-						log!(debug, "💸 Executed offchain worker thread without errors.");
-					}
-				}
+			use offchain_election::{
+				set_check_offchain_execution_status, compute_save_and_submit, submit_queued,
+				OFFCHAIN_REPEAT,
+			};
+			// ensure that we don't run OCW in any case more at least with 5 blocks delay.
+			let threshold: T::BlockNumber = OFFCHAIN_REPEAT.into();
+
+			let election_status = Self::era_election_status();
+			if election_status.is_open_at(now) {
+				// If era election status is open at the current block, mine the solution, save it
+				// and submit it.
+				let initial_output = set_check_offchain_execution_status::<T>(now, threshold)
+					.and_then(|_| compute_save_and_submit::<T>());
+				log!(debug, "initial OCW output at {:?} = {:?}", now, initial_output);
+			} else if election_status.is_open() && Self::queued_elected().is_none() {
+				// If the election window is open, and we don't have a queued solution, if you have
+				// one stored in the OCW DB, the submit it. We again check to never run the OCW more
+				// than every 5 blocks.
+				let resubmit_output = set_check_offchain_execution_status::<T>(now, threshold)
+					.and_then(|_| submit_queued::<T>());
```

Comment:
Two (related) questions:
1. How can you be sure that the queued solution is for the right election status?
2. You have no guarantee that offchain worker runs for every block, so it might skip the `is_open_at(now)` block and you might not have any solutions computed at all.

Given that now we have this extra offchain db entry, why not simplify the code to something like:
```
if elections_status.is_open() && Self::queued_elected().is_none() {
  // 1. acquire a lock
  // 2. compute the solution if not in storage or if era has changed and save it to storage
  // 4. Attempt to submit the solution from storage
}
```

This will make sure that no matter what block OCW is triggered the node will always try to compute a solution if elections are open and there is no other solution submitted yet.

Reply:
That's an interesting point. I will give it a try.

Reply:
Hmm -- okay, code will be much simpler, but the downside is that we want the validators to always run the OCW at least once to be able to topple a potentially weak signed solution. In this scenario if someone manages to get a weak solution into the first block of the election window (i.e. the author of the first block of the election window) no one else will challeneg that.

Reply:
Lastly, I decided to actually keep your recommendation in the sense that the second arm will now try to compute a new solution if it is not there, but I will keep the first arm as well for the reason mentioned. This is actually a bit more complicated than the original thing, but it damn surely should ensure that we are never left without a solution.